### PR TITLE
Avoid using default process group in ProcessGroupAgent.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39916 Enhance DDP docstrings for DDP + RPC support.
* **#39909 Avoid using default process group in ProcessGroupAgent.**

As described in https://github.com/pytorch/pytorch/issues/33583,
ProcessGroupAgent initializes the default process group and this causes issues
if the user initializes the default process group themsleves. Either the RPC
initialization would fail or the user's process group initialization would
fail.

To avoid this, I've changed ProcessGroupAgent init to create its own
ProcessGroupGloo and not use the default one at all.

Closes: https://github.com/pytorch/pytorch/issues/33583

Differential Revision: [D22011868](https://our.internmc.facebook.com/intern/diff/D22011868/)

### CODE REVIEW

File: torch/distributed/rpc/backend_registry.py [hunk 1]
```diff
@@ -106,45 +106,31 @@ def _process_group_init_backend_handler(
...
-        rank=rank,
-        world_size=world_size,
-        timeout=process_group_timeout,
-    )
+    # We're using a bunch of private APIs here since `new_group` requires the
+    # default group to be initialized.
+    group = dist.ProcessGroupGloo(store, rank, world_size, process_group_timeout)
 
-    try:
-        group = dc10d._get_default_group()
```
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py:125]: remove the unused dc10d import?

File: torch/distributed/rpc/backend_registry.py [hunk 2]
```diff
@@ -187,42 +173,27 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
...
-                )
+    if (rank != -1) and (rank != group.rank()):
+        raise RuntimeError(
+            "rank argument {} doesn't match pg rank {}".format(rank, group.rank())
+        )
+    if (world_size != -1) and (world_size != group.size()):
+        raise RuntimeError(
+            "world_size argument {} doesn't match pg size {}".format(
+                world_size, group.size()
             )
```
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py:191]: Does it worth it to dedup these code with the one in PG backend?

**Review by mrshenli — COMMENTED** [review]
This is great! Thanks for adding this fix! We will need to modify some DDP+RPC tests accordingly.

**Comment by mrshenli** [comment]
Test failure looks relevant

```
Jun 13 05:43:17 ======================================================================
Jun 13 05:43:17 ERROR [100.107s]: test_backward_ddp_outside (__main__.TestDdpUnderDistAutograd)
Jun 13 05:43:17 ----------------------------------------------------------------------
Jun 13 05:43:17 Traceback (most recent call last):
Jun 13 05:43:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
Jun 13 05:43:17     self._join_processes(fn)
Jun 13 05:43:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
Jun 13 05:43:17     self._check_return_codes(elapsed_time)
Jun 13 05:43:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 344, in _check_return_codes
Jun 13 05:43:17     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
Jun 13 05:43:17 RuntimeError: Process 0 terminated or timed out after 100.07950329780579 seconds
```

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Comment by pritamdamania87** [comment]
@mrshenli Looks like CI is clean now.

**Pull request closed without merging**